server/src/main/java/org/elasticsearch/index/engine/LiveVersionMap.java

```diff
@@ -139,6 +149,8 @@ Maps buildTransitionMap() {
          * builds a new map that invalidates the old map but maintains the current. This should be called in afterRefresh()
          */
         Maps invalidateOldMap() {
```

Comment:
So I think what we were discussing yesterday was whether it could be possible to move the call of `invalidateOldMap()` from `afterRefresh()` into the `afterFlush()` method? If that would be possible, maybe it's not necessary to do the archiver or the unpromotable refresh service anymore?

Reply:
That would mean that we will accumulate entries that have been evacuated due to local refreshes from the current map and entries that are waiting for a flush to be evacuated from the old map, i.e., we have only one place to track evacuated ids (due to refreshes/flushes). This will make it a bit trickier to prune that old map correctly upon a flush. While a flush is happening (and we're waiting for unpromotable refreshes to make sure flush is visible on the search shards) there could be new refreshes locally on the index shard. Once we get back the flush confirmation, we'd need to go through the map and remove those entries that were propagated in the confirmed flush and leave the refreshed-but-not-flushed yet entries in place. 

I think this way of maintaining the old map (using another map) would make the clean up step more straight forward (and potentially more efficient).

Reply:
> I was thinking refresh/ingestion would be paused during a flush.

There is a flush lock I think, but after flush waiting for the callback for the unpromotable refresh action is async so we need to be able to handle that somehow.

Reply:
Then I think it may still be OK. Since during that time (until the unpromotable refresh is completed) searches can still be served on the promotable.

Reply:
> Could that be done by comparing the translog maxseqno instead? e.g., if it's moved (past what the unpromotable shard has), then we know there are newer ops in the promotable shard and thus the search needs to be done there?

We talked about this. This means if we get (with realtime flag which is true by default) a doc, as long as there has been any changes on that index shard, we'd serve from the indexing shard, which is only slightly better than the current workaround of always serving gets from the indexing shard.